domain/user/state/state.go

```diff
@@ -0,0 +1,395 @@
+// is returned that satisfies usererrors.NotFound.
+func (st *State) SetActivationKey(ctx context.Context, uuid user.UUID, activationKey string) error {
+	err := uuid.Validate()
+	if err != nil {
+		return fmt.Errorf("cannot validate user uuid %q: %w", uuid, err)
+	}
+
+	removed, err := st.isRemoved(ctx, uuid)
+	if err != nil {
+		return fmt.Errorf("cannot get user %q: %w", uuid, err)
+	}
+	if removed {
+		return fmt.Errorf("cannot set activation key for removed user %q", uuid)
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = removePasswordHash(ctx, tx, uuid)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return setActivationKey(ctx, tx, uuid, activationKey)
+	})
+}
+
+// SetPasswordHash removes any active activation keys and sets the user
+// password hash and salt. If no user is found for the supplied name an error
+// is returned that satisfies usererrors.NotFound.
+func (st *State) SetPasswordHash(ctx context.Context, uuid user.UUID, passwordHash string, salt []byte) error {
+	err := uuid.Validate()
+	if err != nil {
+		return fmt.Errorf("cannot validate user uuid %q: %w", uuid, err)
+	}
+
+	removed, err := st.isRemoved(ctx, uuid)
+	if err != nil {
+		return fmt.Errorf("cannot get user %q: %w", uuid, err)
+	}
+	if removed {
+		return fmt.Errorf("cannot set password hash for removed user %q", uuid)
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = removeActivationKey(ctx, tx, uuid)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return setPasswordHash(ctx, tx, uuid, passwordHash, salt)
+	})
+}
+
+func addUser(ctx context.Context, tx *sql.Tx, uuid user.UUID, user user.User, creatorUuid user.UUID) error {
+	addUserQuery := "INSERT INTO user (uuid, name, display_name, created_by_uuid, created_at) VALUES (?, ?, ?, ?, ?)"
+
+	_, err := tx.ExecContext(ctx, addUserQuery,
+		uuid.String(),
+		user.Name,
+		user.DisplayName,
+		creatorUuid.String(),
+		time.Now(),
+	)
+	if databaseutils.IsErrConstraintUnique(err) {
+		return usererrors.AlreadyExists
+	} else if databaseutils.IsErrConstraintForeignKey(err) {
+		return usererrors.UserCreatorUuidNotFound
+	} else if err != nil {
+		return fmt.Errorf("cannot insert user %q: %w", user.Name, err)
+	}
+
+	return nil
+}
+
+func setPasswordHash(ctx context.Context, tx *sql.Tx, uuid user.UUID, passwordHash string, salt []byte) error {
+	defineUserAuthenticationQuery := `
+INSERT INTO user_authentication (user_uuid, disabled) 
+VALUES (?, ?) 
+ON CONFLICT(user_uuid) DO UPDATE SET user_uuid = excluded.user_uuid, disabled = excluded.disabled
+`
+
+	_, err := tx.ExecContext(ctx, defineUserAuthenticationQuery, uuid, false)
+	if err != nil {
+		return fmt.Errorf("cannot set authentification for user %q: %w", uuid, err)
```

Comment:
We seem to be mixing our wording model a lot. Some times we say `user %q` when it is both a username and uuid we erroring on and other times we say `user uuid %q` or `user id %q`. Are we able to update all of the error messages on to a consistent approach?